### DISCUSSION TITLE
SmartNoise SQL Support for all SQL dialects

### DISCUSSION THREAD

**Discussion by oskarkocol** [opendp/opendp#733] [SmartNoise] [answered]
In theory, is there anything stopping us from supporting all SQL dialects? Instead of building our own SQL parsers, could we rely on an external tool like [SQLGlot](https://github.com/tobymao/sqlglot)?

I worked some time ago on bringing support for the BigQuery SQL dialect into SmartNoise SQL. It still seems to me that bringing DP-SQL to popular data warehousing providers is key for broader adoption of DP. But BigQuery alone is just a drop in a bucket. Redshift seems like an obvious next step, but then, why not Snowflake or Databricks? On other hand, [duckdb](https://github.com/duckdb/duckdb) has been gaining on popularity recently, and I foresee more SQL dialects to come as data analytics tools are being built.

One way that could work:
1. to have an **external** transpiler building an AST
2. to have smartnoise-sql manipulate such AST into a differentially private AST
3. to use an external compiler to bring the AST back to SQL
 
The obvious project that I have on my mind is SQLGlot, which as of today supports [16 SQL dialects](https://github.com/tobymao/sqlglot/blob/7d98790e8b1ced09b714a408209a2a3ad48520cc/sqlglot/dialects/__init__.py#L59-L79)

Of course, taking on an additional dependency is a big decision. I can see the following candidates:
* [SQLGlot](https://github.com/tobymao/sqlglot)(3.2k stars)
* [sqlfluff](https://github.com/sqlfluff/sqlfluff) (6.1k stars)
* [sqltree](https://github.com/JelleZijlstra/sqltree) (27 stars)
* [sqlparse](https://github.com/andialbrecht/sqlparse) (3.2k stars)
* ~~[moz_sql_parser](https://github.com/mozilla/moz-sql-parser) (500 stars)~~ Deprecated.
* [sqloxide](https://github.com/wseaton/sqloxide) (88 stars, Python bindings for [sqlparser-rs](https://github.com/sqlparser-rs/sqlparser-rs) 1.8k stars)

All of these packages are [benchmarked here](https://github.com/tobymao/sqlglot#benchmarks)

**Comment by Shoeboxam** [comment]
Tagging @joshua-oss

**Comment by joshua-oss** [comment, marked as answer]
This is a great idea.  Limiting ourselves to SQL-92 is making it painful to support (e.g.) MIN and MAX, because every engine has different syntax for how to return a sampled list of values in a list inside an aggregate.  Implementing MIN and MAX therefore requires manually rewriting of code for essentially every engine.

My hope was that SQLAlchemy would solve that problem for us, since it allows creation of code that can target many engines.  However, it doesn't seem to be able to map from a single object model to all engines.  As far as I can tell, the two examples below would require different Python code in SQLAlchemy to generate:

# MySQL: Select 50 random ages, from each gender

```sql
WITH numbered_ages AS (
  SELECT 
    sex, 
    age,
    ROW_NUMBER() OVER(PARTITION BY sex ORDER BY RAND()) as rn
  FROM pums
)
SELECT 
  sex, 
  GROUP_CONCAT(age ORDER BY rn SEPARATOR ',') as ages
FROM numbered_ages
WHERE rn <= 50
GROUP BY sex;
```

# SqlServer: Select 50 random ages, from each gender

```sql
WITH numbered_ages AS (
  SELECT 
    sex, 
    age,
    ROW_NUMBER() OVER(PARTITION BY sex ORDER BY NEWID()) as rn
  FROM pums
)
SELECT 
  sex, 
  STRING_AGG(age, ',') WITHIN GROUP (ORDER BY rn) as ages
FROM numbered_ages
WHERE rn <= 50
GROUP BY sex;
```

These examples differ in both the random function and the syntax of `STRING_AGG` vs. `GROUP_CONCAT`.  FWIW, I had GPT-4 generate these, so there may be slight errors.  The point being that there is no universal way to do this in SQLAlchemy.  It may only be possible with a transpiler.  We could of course hand-code mappings in every `RandomEngineReader`, but this is already cumbersome for the language customizations we currently support.

One way to evaluate this decisions might be to try this example and a few other that might be useful (e.g. subqueries and joins), to see which one of the libraries above solve the problem without requiring us to write a bunch of engine-specific overrides.  It might also be the case that some engines have no language syntax that can be targeted by a transpiler for a given task, and it would be good if the library can tell us that, so the feature could be disabled, or some other workaround applied.

The engine-specific overrides are not necessarily the hardest part of supporting new engines.  Equally as painful is the process of setting up reliable test environments for new feature development and CI/CD.  That has been the main blocker to supporting Snowflake.  Having a very good transpiler might make us feel more confident about changing things, but we probably still need a safety net in the form of provisioned test environments.  This gets more painful as we add new test databases.  We would really like to add a test database that includes multiple relational tables in 3NF, a test database showing extremely high-dimensional sequence data like click logs or ad tracking, and one that shows high-dimensional and sparse data where the dimensionality comes from the number of columns (e.g. each column may have cardinality of < 10,000 and most mass in < 500 discrete values, but the combined cardinality across columns is like 10^30).  The pain of setting up these test databases for all engines is a blocker.  SQLAlchemy promised to help with DDL and loading from CSV across multiple engines, but in practice it is buggy and way too slow.  Proprietary bulk insert is a better choice for most engines.  It would be incredible if there were a library that could transpile the DDL and bulk insert scripts (including handling of NULL, etc.).

Another area where we are currently challenged to generate cross-engine code is with the suite of evaluation metrics and membership inference attacks we would like to add to the library.  We currently delegate to the engine just the portion of the job that involves creating the aggregations for differential privacy, and all other work is assumed to be done outside the engine.  However, post-processing to compute the average error across 100 runs, or computing the ability to fingerprint individuals, are quite a bit more computationally expensive than the actual mitigation.  For toy examples like PUMS, this is no big deal, but for real-world examples that may include many billions of rows grouped into tens of millions of bins, we need the power of an engine. At the moment, it seems easiest to just require all of this post-processing to be done in Spark (and that is what we do), but it would be ideal if we could be engine-agnostic.

Note that the current limited approach is a conscious design decision we made in [1], described under "Implementation Approach".  However, it seems we are reaching the limits of the "Manipulate SQL" approach and we need to support some slightly more general intermediate representation.  Opening up to an IR makes it more difficult to ensure correctness and security.  We currently control everything from lexer to grammar to AST, so we would be changing to place a lot of trust in an external library for protection against injection attacks and privacy violations and so on.  I don't know how clear these transpilers are when there are differences in semantics of the transpiled code, or how we would be confident that all of our assumptions would hold.  I notice there are no CI/CD tests in SQLGlot for the various engines they support, so it seems sort of "best effort".

An additional thought: I've held out hope that we could replace all of our parsing and rewriting code with Rust instead of Python.  We have differentially private workloads running in ACC and TDX that can use Python, but there are workloads where the TCB must be absolutely minimized, and that means SGX.  We can run OpenDP in attested Rust SGX enclaves, but we lose all ability to do interactive query.  This would be very easy if the building blocks for the AST/rewriter/transpiler were written in Rust instead of Python.  Longer term, these security-sensitive and privacy-sensitive interop layers might make most sense as Rust crates.

[1] https://github.com/opendp/smartnoise-sdk/blob/main/sql/docs/papers/Design%20Approach%20-%20SQL.pdf
